samples/extensions/khr/conway/main.cpp

```diff
@@ -0,0 +1,366 @@
+            std::cerr << std::string(shader_log.cbegin(), shader_log.cend()) << std::endl;
+        }
+
+        return shader;
+    };
+    auto create_program = [](std::initializer_list<GLuint> shader_stages)
+    {
+        GLuint program = glCreateProgram(); checkError("glCreateProgram()");
+        for (auto shader_stage : shader_stages) {
+            glAttachShader(program, shader_stage); checkError("glAttachShader(program, shader_stage)"); }
+
+        glLinkProgram(program); checkError("glLinkProgram(program)");
+        GLint status = GL_FALSE;
+        glGetProgramiv(program, GL_LINK_STATUS, &status); checkError("glGetProgramiv(program, GL_LINK_STATUS, &status)");
+        if(status != GL_TRUE)
+        {
+            int log_length = 0;
+            glGetProgramiv(program, GL_INFO_LOG_LENGTH, &log_length); checkError("glGetProgramiv(program, GL_INFO_LOG_LENGTH, &log_length)");
+            std::vector<GLchar> program_log(log_length);
+            glGetProgramInfoLog(program, log_length, NULL, program_log.data()); checkError("glGetProgramInfoLog(program, log_length, NULL, program_log.data())");
+            std::cerr << std::string(program_log.cbegin(), program_log.cend()) << std::endl;
+        }
+
+        return program;
+    };
+
+    vertex_shader   = create_shader("./conway.vert.glsl", GL_VERTEX_SHADER);
+    fragment_shader = create_shader("./conway.frag.glsl", GL_FRAGMENT_SHADER);
+    gl_program = create_program({ vertex_shader, fragment_shader });
+
+    std::vector<float> quad =
+        //  vertices  , tex coords
+        //  x  ,   y  ,  u  ,   v
+        { -1.0f, -1.0f, 0.0f, 0.0f,
+          -1.0f,  1.0f, 0.0f, 1.0f,
+           1.0f, -1.0f, 1.0f, 0.0f,
+           1.0f,  1.0f, 1.0f, 1.0f };
+
+    glGenBuffers(1, &vertex_buffer); checkError("glGenBuffers(1, &vertex_buffer)");
+    glBindBuffer(GL_ARRAY_BUFFER, vertex_buffer); checkError("glBindBuffer(GL_ARRAY_BUFFER, vertex_buffer)");
+        glBufferData(GL_ARRAY_BUFFER, quad.size() * sizeof(float), quad.data(), GL_STATIC_DRAW); checkError("glBufferData(GL_ARRAY_BUFFER, quad.size() * sizeof(float), quad.data(), GL_STATIC_DRAW)");
+    glBindBuffer(GL_ARRAY_BUFFER, 0); checkError("glBindBuffer(GL_ARRAY_BUFFER, 0)");
+
+    glGenVertexArrays(1, &vertex_array); checkError("glGenVertexArrays(1, &vertex_array)");
+    glBindVertexArray(vertex_array); checkError("glBindVertexArray(vertex_array)");
+        glBindBuffer(GL_ARRAY_BUFFER, vertex_buffer); checkError("glBindBuffer(GL_ARRAY_BUFFER, vertex_buffer)");
+        glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL)); checkError("glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL))");
+        glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL + 2 * sizeof(float))); checkError("glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL + 2 * sizeof(float)))");
```

Comment:
I'm getting a warning here too:

```
/home/bashbaug/git/OpenCL-SDK/samples/extensions/khr/conway/main.cpp: In member function ‘virtual void Conway::initializeGL()’:
/home/bashbaug/git/OpenCL-SDK/samples/extensions/khr/conway/main.cpp:198:108: warning: NULL used in arithmetic [-Wpointer-arith]
  198 |         glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL + 2 * sizeof(float))); checkError("glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 4*sizeof(float), (GLvoid *)(NULL + 2 * sizeof(float)))");
      |  
```

Reply:
My colleague caught that one too. I'll change it to 0 to silence the warning. The OpenGL spec explicitly says that the 0 pointer is the beginning of the buffer.

> Specifies a offset of the first component of the first generic vertex attribute in the array in the data store of the buffer currently bound to the GL_ARRAY_BUFFER target. The initial value is 0.

Because it's a pointer, and the value is 0, I wanted to make it explicit by using `NULL` but the warning triggers.

Reply:
Fixed.